src/bookshelf/src/bksw_sec.erl

```diff
@@ -20,128 +21,156 @@
-                                              Expires,
-                                              Headers)
-    catch
-        _:Why ->
-            error_logger:error_report({error, {{mini_s3, make_signed_url_authorization},
-                                               [<<"SECRETKEY">>, Method, Path, Expires, Headers],
-                                               Why}}),
-            error
-    end.
+    % If the signature comparison fails, we attempt an alternate signature
+    % calculation and comparison which adds or removes the port from the host
+    % header depending on whether the port is present or absent. At the time
+    % this was put in, there were problems with signature failures due to
+    % inconsistent treatment of host headers by various chef clients
+    % (present or missing ports).  Determining whether alt sig comparison is
+    % still necessary would require investigation and testing.
 
-do_standard_authorization(RequestId, IncomingAuth, Req0, Context) ->
-    Headers = mochiweb_headers:to_list(wrq:req_headers(Req0)),
-    AmzHeaders = amz_headers(Headers),
-    RawMethod = wrq:method(Req0),
-    Method = string:to_lower(erlang:atom_to_list(RawMethod)),
-    ContentMD5 = proplists:get_value('Content-Md5', Headers, ""),
-    ContentType = proplists:get_value('Content-Type', Headers, ""),
-    Date = proplists:get_value('Date', Headers, ""),
-    %% get_object_and_bucket decodes the bucket, but the request will have been signed with
-    %% the encoded bucket.
-    {ok, Bucket0, Resource} = bksw_util:get_object_and_bucket(Req0),
-    Bucket = bksw_io_names:encode(Bucket0),
-    AccessKey = bksw_conf:access_key_id(Context),
-    SecretKey = bksw_conf:secret_access_key(Context),
-
-    {StringToSign, RawCheckedAuth} =
-        mini_s3:make_authorization(AccessKey, SecretKey,
-                                   erlang:list_to_existing_atom(Method),
-                                   bksw_util:to_string(ContentMD5),
-                                   bksw_util:to_string(ContentType),
-                                   bksw_util:to_string(Date),
-                                   AmzHeaders, bksw_util:to_string(Bucket),
-                                   "/" ++ bksw_util:to_string(Resource),
-                                   ""),
-    CheckedAuth = erlang:iolist_to_binary(RawCheckedAuth),
-    [AccessKeyId, Signature] = split_authorization(IncomingAuth),
-    case erlang:iolist_to_binary(IncomingAuth) of
-        CheckedAuth ->
-            {true, Req0, Context};
+    CalculatedSig =
+        case const_time_compare(IncomingSignature, ComparisonSig, true) of
+            true ->
+                IncomingSignature;
+            _ ->
+                AltSigV4Headers   = erlcloud_aws:sign_v4(method(Auth), path(Auth), config(Auth), alt_signed_headers(Auth), <<>>, Region, "s3", QueryParams, Date),
+                _AltComparisonSig = parseauth(proplists:get_value("Authorization", AltSigV4Headers, ""))
```

Comment:
We can skip the variable assignment here - just invoking `parseauth` should assign that return as the value of `CalculatedSig`

Reply:
Done.